test.py: use internal id to manage servers

alecco opened the pull request:
Instead of using assigned IP addresses, use a local integer ID for managing servers. IP address can be reused by a different server.

While there, get host ID (UUID). This can also be reused with `node replace` so it's not good enough for tracking.

alecco commented:
Changes in v2  (https://github.com/scylladb/scylladb/commit/9a8441a51b2deffc1d24133a04bda4628afb2f03):

- return ServerInfo for adding server
- fix missing places using `str` instead of `HostID` type
- add a few asserts on json returns

alecco commented:
Forgot to say, this pr consists of 4 commits and it's based on top of https://github.com/scylladb/scylladb/pull/11725/

alecco commented:
Changes in v3 https://github.com/scylladb/scylladb/commit/bb1065b12dee013ace1bf7d909775ac2bbe981bc:

- Fix `test_raft_upgrade` IPs

kostja commented:
The patch is LGTM, two comments:
- suggest to extract renames into a separate patch (hostname -> host_ip)
- suggest to consistently use 'ip_addr' to designate an ip address. the current patch uses ip, ip_, host_ip - different names. Let's stick to one, ip_addr is widespread in scylla core source code.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2602/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/6576effceebc5378c15dedf597d1d11abbbf26ec:

- Separate `ScyllaServer.hostname` change to its own commit
- Use `host_id` and `host_ip` instead of `id_` and `ip`
- Minor logging improvements
- minor pylint fixes

nyh commented:
Can you please add to the cover letter a few words about the motivation of this patch? What was wrong with the IP address - is it about tests which try to change a node's IP address?

kbr- reviewed (commented):
I don't think Host IDs are the right tool for the job...

The problem is node replace, where a new node takes over an existing node's Host ID. So at some point a given Host ID will start pointing to a different node, and we won't be able to identify the existing node.

I think what we should do instead is to have our own identifiers in the test framework.

BTW. For the future, I think we should discuss big changes like this one and reach consensus before any code is written.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2612/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2633/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/60700596b577d8f58c71c0c56104e5e2cd468ed2:

- Fix `_cluster_remove_node` to pick `server_id` from JSON args
- Fix missing renames `host_id` to `server_id`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2636/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2654/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/991d1f226dd01c867b9c9b0b72e7cc7ea64b274a:

- workaround on `rest_api_is_up` (will need a better fix, but later)
- rebase on next

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2661/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/bd833f90e5464854290efaa8797cdce178ca50d0 / https://github.com/scylladb/scylladb/commit/bde1d3fee018f1669a7c3eccc1b0e8534adc56e0 (rebase):

- Don't use `aiohttp.ClientSession`, de-duplicate code. (fixes issue with `rest_api/test_task_manager`)
- Improve types, lint, and do sanity checks

(Please ignore https://github.com/scylladb/scylladb/commit/991d1f226dd01c867b9c9b0b72e7cc7ea64b274a)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2693/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/f77b6c465651d1dc0e3453782cc80852bf68df3b:

- removed `get_json` type check

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2696/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2697/

nyh reviewed (changes requested):
Looks mostly good, but I left a few relatively minor complaints.

However, one overarching comment I have (that doesn't necessarily need to be solved in this PR, if you don't want to) is that there is too much unnecessary complexity here... After this patch you have at least three separate concepts of node IDs - you have an IP address, a UUID, and a numeric index. In addition to these three, you have multiple "wrappers" of these basic types into several more complex Python classes. And you use all these different types and wrappers somewhere in the code and I find it very hard to follow, and unnecessary:

Personally, I would choose just one identifier - namely, the numeric index - and use it exclusively in the code. You can then have a function to return node i's object, and this object contain this node's IP address if you need it. No function would ever need to take the IP address directly. The third concept, the "host uuid" doesn't need to be be central to the API at all - it can just be a function to call the REST API and retrieve this "host uuid", when (and if) any caller needs it. For efficiency, the result of this function can be cached in case it is called several times for the same node to avoid calling the REST API every time. But it doesn't need to be a central part of this API, and doesn't need to be set right in the node setup function.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -291,19 +292,21 @@ async def cql_is_up(self) -> bool:
-                    return True
-        except aiohttp.ClientConnectionError:
+                    pass
+                self.host_id = await api.get_host_id(self.hostname)
```
I missed deleting the old code in the refactor.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/eae1194fa63f0f18ba03461ec79586c91741e4fc

kbr- commented:
Please don't use the name "Host IP" - it's visually similar to "Host ID", and in Scylla codebase we usually just say "addr".
So how about "ip_addr" - like Kostja suggested in an earlier review?

kbr- commented on test/pylib/manager_client.py:
```diff
@@ -104,70 +118,91 @@ async def replicas(self) -> int:
-        host_list = await self.client.get_text("/cluster/running-servers")
-        return host_list.split(",")
+    async def running_servers(self) -> List[ServerInfo]:
+        """Get dict of host id to IP address of running servers"""
```
Please do a self-review and fix all comments, like this one, to use the proper concepts from the latest version of this PR.
For example here: it is neither a dict, nor does it use Host IDs.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/eae1194fa63f0f18ba03461ec79586c91741e4fc

kbr- commented on test/pylib/manager_client.py:
```diff
@@ -104,70 +118,91 @@ async def replicas(self) -> int:
+    async def get_host_ip(self, server_id: ServerID) -> IPAddress:
+        """Get host id through Scylla REST API"""
+        try:
+            host_ip = await self.client.get_text(f"/cluster/host-id/{server_id}")
```
The PR requires a self-review by @alecco. There are lots of places like this one which mix up terminology from before-PR, previous versions of this PR, and the latest version. And the entire thing becomes very confusing.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/eae1194fa63f0f18ba03461ec79586c91741e4fc

kbr- commented on test/pylib/rest_client.py:
```diff
@@ -118,26 +121,29 @@ class ScyllaRESTAPIClient():
         self.client = TCPRESTClient(port)
 
-    async def get_host_id(self, server_id: str) -> str:
+    async def get_host_id(self, server_id: IPAddress) -> HostID:
```
The PR requires a self-review by @alecco. There are lots of places like this one which mix up terminology from before-PR, previous versions of this PR, and the latest version. And the entire thing becomes very confusing.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/eae1194fa63f0f18ba03461ec79586c91741e4fc

kbr- commented on test/pylib/manager_client.py:
```diff
@@ -19,6 +19,19 @@
 
+ServerID = NewType('ServerID', int)
+IPAddress = NewType('IPAddress', str)
+HostID = NewType('HostID', str)
```
The PR requires a self-review by @alecco. There are lots of places like this one which mix up terminology from before-PR, previous versions of this PR, and the latest version. And the entire thing becomes very confusing.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/eae1194fa63f0f18ba03461ec79586c91741e4fc

kbr- commented:
> No function would ever need to take the IP address directly.

Unfortunately the driver requires IP addresses.
So we have this:
1. test framework APIs are operating in terms of `ServerID`s - this is because we must handle the possibility (in the future) of changing IPs of servers (and we cannot use `HostID`s to identify servers because `HostIDs` can be stolen during replacenode).
2. the driver is operating in terms of IP addresses

so the question is how we deal with that - do we fetch the last known IP address from the framework every time we want to use the driver? Or do we store (ServerID, IP addr) pair on the test side - like the PR does IIUC - and we refresh the IP addr when we know that the IP changes (because it can only happen by explicit test operation)? I think we should do the latter, there's no point in fetching the IP every time we want to perform a CQL op.

alecco commented:
> Looks mostly good, but I left a few relatively minor complaints.
> 
> However, one overarching comment I have (that doesn't necessarily need to be solved in this PR, if you don't want to) is that there is too much unnecessary complexity here... After this patch you have at least three separate concepts of node IDs - you have an IP address, a UUID, and a numeric index. In addition to these three, you have multiple "wrappers" of these basic types into several more complex Python classes. And you use all these different types and wrappers somewhere in the code and I find it very hard to follow, and unnecessary:
> 
> Personally, I would choose just one identifier - namely, the numeric index - and use it exclusively in the code. You can then have a function to return node i's object, and this object contain this node's IP address if you need it. No function would ever need to take the IP address directly. The third concept, the "host uuid" doesn't need to be be central to the API at all - it can just be a function to call the REST API and retrieve this "host uuid", when (and if) any caller needs it. For efficiency, the result of this function can be cached in case it is called several times for the same node to avoid calling the REST API every time. But it doesn't need to be a central part of this API, and doesn't need to be set right in the node setup function.

There is only 1 identifier in the code, local integer id. The IP and host ID are stored as members of `ScyllaServer` as they are used frequently and querying for them every time doesn't make sense, and the code will track in the future if either changes (e.g. node replace).

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/eae1194fa63f0f18ba03461ec79586c91741e4fc:

- `ServerID` renamed `ServerNum`
- `host_ip` renamed `host_id`
- Fixed comments
- Fixed commit messages
- Fixed log string
- Fixed `get_host_id()` and `get_host_ip()` to call the right API
- Fixed types of return for `_fetch()` and `get_json()`
- Removed old REST API up check and just use `get_host_id()`
- Check `host_id` set to avoid calling over and over until `CQL` is up

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/be3d9bac7290cc692578c84b28e2a47fdad5e6bb:

- Fix `get_host_id()` comment

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2710/

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2710/

Unrelated dtest failure

alecco commented:
https://github.com/scylladb/scylladb/commit/345f902f83ca062e19a57f3262347afd8f877439 is just a rebase on latest `master` to force new CI run.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2714/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/eca7bb8f4d00285f85781a43727e6be2fc8ea608:

- Fixed `.read_log()`

Changes in https://github.com/scylladb/scylladb/commit/6635e6163463bf74df6826ee15735d79e6493578:

- rebased to latest master

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2778/

kbr- commented on test/pylib/scylla_cluster.py:
```diff
@@ -291,19 +292,17 @@ async def cql_is_up(self) -> bool:
-        except aiohttp.ClientConnectionError:
+            self.host_id = await api.get_host_id(self.hostname)
+            return True
+        except (aiohttp.ClientConnectionError, RuntimeError):
```
Doesn't `RuntimeError` indicate a problem? Shouldn't we propagate it?

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847 https://github.com/scylladb/scylladb/compare/6635e6163463bf74df6826ee15735d79e6493578..0ebda403004c746f8a0019098e9125c160433847

Now `aiohttp.request` raises (its own Exceptions) and we don't do manual re-raising with `RuntimeError`s.

kbr- commented on test/pylib/scylla_cluster.py:
```diff
@@ -291,19 +292,17 @@ async def cql_is_up(self) -> bool:
 
-    async def rest_api_is_up(self) -> bool:
+    async def get_host_id(self, api: ScyllaRESTAPIClient) -> bool:
         """Test that the Scylla REST API is serving. Can be used as a
```
Docstring needs updating.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847 https://github.com/scylladb/scylladb/compare/6635e6163463bf74df6826ee15735d79e6493578..0ebda403004c746f8a0019098e9125c160433847

kbr- commented on test/pylib/manager_client.py:
```diff
@@ -19,6 +19,19 @@
+
+
+class ServerInfo(NamedTuple):
+    """Server id (UUID) and IP address"""
```
"Server id (UUID)"

alecco replied:
The named tuple has the IP address also

kbr- replied:
But it doesn't have a "Server id (UUID)" <- it's an error

kbr- commented on test/pylib/manager_client.py:
```diff
@@ -104,70 +118,91 @@ async def replicas(self) -> int:
-        host_list = await self.client.get_text("/cluster/running-servers")
-        return host_list.split(",")
+    async def running_servers(self) -> List[ServerInfo]:
+        """Get dict of host id to IP address of running servers"""
```
the comment still mentions `dict` and `host id`, which `ServerInfo` does not contain.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847 https://github.com/scylladb/scylladb/compare/6635e6163463bf74df6826ee15735d79e6493578..0ebda403004c746f8a0019098e9125c160433847

kbr- commented on test/pylib/rest_client.py:
```diff
@@ -119,27 +122,29 @@ class ScyllaRESTAPIClient():
                 f"get_host_id: invalid {host_uuid}"
         host_uuid = host_uuid.lstrip('"').rstrip('"')
-        return host_uuid
+        return cast("HostID", host_uuid)
```
You're using `HostID(...)` in a similar situation elsewhere, but here you use `cast("HostID", ...)` - wouldn't `HostID(...)` work as well?

alecco replied:
As above, circular import. This `cast` is understood by type checkers an ignored by the interpreter.

alecco replied:
Fixed on https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847

alecco commented on test/pylib/rest_client.py:
```diff
@@ -5,14 +5,17 @@
 from aiohttp import request, BaseConnector, UnixConnector
 import pytest
+if TYPE_CHECKING:
+    from test.pylib.manager_client import HostID, IPAddress
```
This avoids a circular import.

alecco commented on test/topology_raft_disabled/test_raft_upgrade.py:
```diff
@@ -34,26 +34,26 @@ async def reconnect_driver(manager: ManagerClient) -> Session:
-    await manager.server_start(srv)
-    logging.info(f"{srv} restarted")
+async def restart(manager: ManagerClient, server_id: ServerNum) -> None:
+    logging.info(f"Stopping {server_id} gracefully")
```
Fixed in https://github.com/scylladb/scylladb/commit/7baab82e5c884194b88771b3d0d8b49237c4dc05

alecco commented on test/topology_raft_disabled/test_raft_upgrade.py:
```diff
@@ -76,9 +76,10 @@ async def cql_ready():
-async def wait_for_cql_and_get_hosts(cql: Session, ips: list[str], deadline: float) -> list[Host]:
+async def wait_for_cql_and_get_hosts(cql: Session, servers: list[ServerInfo], deadline: float) \
+        -> list[Host]:
     """Wait until every ip in `ips` is available through `cql` and translate `ips` to `Host`s."""
```
Fixed in https://github.com/scylladb/scylladb/commit/7baab82e5c884194b88771b3d0d8b49237c4dc05

alecco commented on test/pylib/manager_client.py:
```diff
@@ -104,70 +118,91 @@ async def replicas(self) -> int:
+        except Exception as exc:
+            raise Exception(f"Failed to add server") from exc
+        try:
+            s_info = ServerInfo(ServerNum(int(server_info["server_id"])), server_info["ip_addr"])
```
Fixed in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/a7ecac4785cad07eef2dd0d95dfb34dce332afab

alecco commented on test/pylib/manager_client.py:
```diff
@@ -104,70 +118,91 @@ async def replicas(self) -> int:
+        except RuntimeError as exc:
+            raise Exception(f"Failed to get list of running servers") from exc
+        assert isinstance(server_info_list, list), "running_servers got unknown data type"
+        return [ServerInfo(ServerNum(int(info[0])), info[1]) for info in server_info_list]
```
Fixed in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847 https://github.com/scylladb/scylladb/compare/6635e6163463bf74df6826ee15735d79e6493578..0ebda403004c746f8a0019098e9125c160433847

alecco commented on test/pylib/rest_client.py:
```diff
@@ -5,132 +5,112 @@
+                return await getattr(resp, response_type)()
+        return None  # Unreached
+
+    async def get(self, resource_uri: str, host: Optional[str] = None, port: Optional[int] = None,
```
Removed json arg in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847 https://github.com/scylladb/scylladb/compare/6635e6163463bf74df6826ee15735d79e6493578..0ebda403004c746f8a0019098e9125c160433847

Port feature left as it's already in and it's a hassle to remove it just because. It's custom to support custom port in client code.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/0ebda403004c746f8a0019098e9125c160433847 (https://github.com/scylladb/scylladb/compare/6635e6163463bf74df6826ee15735d79e6493578..0ebda403004c746f8a0019098e9125c160433847):

- Fix manager stop on error on cluster startup
- Make `aiohttp.request` raise exception instead of manually checking `status`
- Fix docstrings
- Fix missing `IPAddress` cast
- Remove useless f-strings
- Raise `from` exception when there was one

Changes in https://github.com/scylladb/scylladb/commit/7baab82e5c884194b88771b3d0d8b49237c4dc05 (https://github.com/scylladb/scylladb/compare/0ebda403004c746f8a0019098e9125c160433847..7baab82e5c884194b88771b3d0d8b49237c4dc05):

- Fix docstring

Changes in https://github.com/scylladb/scylladb/commit/379a3d5d33c3b9b2ac1608fcc7744373471eea3e (https://github.com/scylladb/scylladb/compare/7baab82e5c884194b88771b3d0d8b49237c4dc05..379a3d5d33c3b9b2ac1608fcc7744373471eea3e):

- Rebase to latest master 

Changes in https://github.com/scylladb/scylladb/commit/a7ecac4785cad07eef2dd0d95dfb34dce332afab (https://github.com/scylladb/scylladb/compare/379a3d5d33c3b9b2ac1608fcc7744373471eea3e..a7ecac4785cad07eef2dd0d95dfb34dce332afab):

- Fixed missing cast

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2827/

kbr- commented:
@alecco it seems that your PR grew bigger since I reviewed it 2 days ago. You added another commit with additional bunch of fixes which IIUC are not related to the review.
I noticed this happened in some of your other PRs too - the scope of PR growing while it is being reviewed.

Please try to avoid this. If you find additional issues you want to fix, don't include them into an existing PR unless the issues are introduced by that PR (but the cluster startup failure issue was a preexisting one AFAIK). Doing that confuses the reviewers, makes the process longer and delays merging the PR,

alecco commented:
> @alecco it seems that your PR grew bigger since I reviewed it 2 days ago. You added another commit with additional bunch of fixes which IIUC are not related to the review.

Only one trivial commit in the beginning testing for existence of members.

The other added extra commits came from your reviews mostly.

kbr- commented:
> Only one trivial commit in the beginning testing for existence of members.

It's not trivial and it contains bugs.

I'll review it although I'd prefer for it not to be added to the PR at all.

kbr- reviewed (changes requested):
I had a hunch and I checked how node op failures appear in the logs before and after this PR.
I injected a failure into the removenode operation:
```diff
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 03611a7ed0..b04ebe0e45 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -2347,6 +2347,8 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
                 ignore_nodes.push_back(hoep.endpoint);
             }
 
+            throw std::runtime_error{"Failing removenode because of reasons"};
+
             // Step 1: Decide who needs to sync data
             //
             // By default, we require all nodes in the cluster to participate
```

and ran the test `test_topology.py::test_remove_node_add_column`.

**Before this PR** we would see the following.
In `test_topology.1.log` we have:
```
>           raise RuntimeError(f"status code: {resp.status}, body text: {text}, "
                               f"resource {resource_uri} json {json}")
E           RuntimeError: status code: 500, body text: Error removing 127.88.75.2 68014507-6479-4a32-a188-25af216b394f status code: 500, body text: {"message": "std::runtime_error (Failing removenode because of reasons)", "code": 500}, resource http://127.88.75.1:10000/storage_service/remove_node params {'host_id': '
68014507-6479-4a32-a188-25af216b394f', 'ignore_nodes': ''}, resource http+unix://api/cluster/remove-node/127.88.75.1 json {'to_remove_ip': '127.88.75.2', 'to_remove_host_id': '68014507-6479-4a32-a188-25af216b394f', 'ignore_dead': []}
```

very nice, I can see what the original reason of failure was (the exact exception thrown by Scylla).
In `test.py.dev.log`:
```
12:11:08.359 ERROR> _cluster_remove_node failed initiator 127.88.75.1 server 127.88.75.2 68014507-6479-4a32-a188-25af216b394f ignore_dead [], check log at /home/kbraun/dev/scylla/testlog/dev/scylla-1.log
```
it doesn't provide much details, but at least it points to the node's logs. Unfortunately, node's logs don't provide details in this case (Scylla only threw an exception which propagated to the client).

**After this PR** we have the following.
In `test_topology.1.log` we have:
```
>       await manager.remove_node(servers[0].server_id, servers[1].server_id)   # Remove   [1]

test/topology/test_topology.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/pylib/manager_client.py:179: in remove_node
    await self.client.put_json(f"/cluster/remove-node/{initiator_id}", data)
test/pylib/rest_client.py:85: in put_json
    await self._fetch("PUT", resource_uri, host = host, port = port, params = params,
test/pylib/rest_client.py:49: in _fetch
    async with request(method, uri,
/usr/lib64/python3.10/site-packages/aiohttp/client.py:1186: in __aenter__
    self._resp = await self._coro
/usr/lib64/python3.10/site-packages/aiohttp/client.py:640: in _request
    resp.raise_for_status()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ClientResponse(http+unix://api/cluster/remove-node/1) [500 Internal Server Error]>
<CIMultiDictProxy('Content-Type': ...rset=utf-8', 'Content-Length': '247', 'Date': 'Wed, 02 Nov 2022 11:15:03 GMT', 'Server': 'Python/3.10 aiohttp/3.8.1')>


    def raise_for_status(self) -> None:
        if not self.ok:
            # reason should always be not None for a started response
            assert self.reason is not None
            self.release()
>           raise ClientResponseError(
                self.request_info,
                self.history,
                status=self.status,
                message=self.reason,
                headers=self.headers,
            )
E           aiohttp.client_exceptions.ClientResponseError: 500, message='Internal Server Error', url=URL('http+unix://api/cluster/remove-node/1')
```
which is pretty much useless.
in `test.py.dev.log` we have the same as before
```
12:15:03.604 ERROR> _cluster_remove_node failed initiator ScyllaServer(1, 127.106.139.1, a9a37b7c-a5a1-4676-ac38-0e46676290e5) server ScyllaServer(2, 127.106.139.2, 4f871612-22b1-4aa1-b3a9-798fc73db673) ignore_dead [], check log at /home/kbraun/dev/scylla/testlog/dev/scylla-1.log
```
but the node's logs as before don't provide the details.

**We lost the cause of the failure** - the message from the exception.
This is a huge usability regression. The logs are now useless.

When performing refactors like this, we must check if they don't break our logs.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -743,9 +743,6 @@ async def stop(self) -> None:
             del self.site
-        if hasattr(self, "api"):
-            await self.api.close()
-            del self.api
```
No session, no need to await `close()`.

alecco replied:
Removed in https://github.com/scylladb/scylladb/commit/325cbef7d4f53d50b37fd4f7421da0ceb89f37bd

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/325cbef7d4f53d50b37fd4f7421da0ceb89f37bd (https://github.com/scylladb/scylladb/compare/a7ecac4785cad07eef2dd0d95dfb34dce332afab..325cbef7d4f53d50b37fd4f7421da0ceb89f37bd):

- Removed unrelated commit
- Reverted REST client to raise custom exception with message
- Fixed docstring

Changes in https://github.com/scylladb/scylladb/commit/20cd5a222e02c421936cc94388dda80f73d2be6e (https://github.com/scylladb/scylladb/compare/325cbef7d4f53d50b37fd4f7421da0ceb89f37bd..20cd5a222e02c421936cc94388dda80f73d2be6e):

- Rebase to latest master

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -740,17 +740,22 @@ async def _before_test(self, test_case_name: str) -> None:
     async def stop(self) -> None:
```
Commit removed in https://github.com/scylladb/scylladb/commit/325cbef7d4f53d50b37fd4f7421da0ceb89f37bd

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -740,17 +740,22 @@ async def _before_test(self, test_case_name: str) -> None:
+        if hasattr(self, "api"):
+            await self.api.close()
+            del self.api
+        if hasattr(self, "api"):
```
Removed in https://github.com/scylladb/scylladb/commit/325cbef7d4f53d50b37fd4f7421da0ceb89f37bd

alecco commented:
> I had a hunch and I checked how node op failures appear in the logs before and after this PR. I injected a failure into the removenode operation:
> [...]
> When performing refactors like this, we must check if they don't break our logs.

Fixed in https://github.com/scylladb/scylladb/commit/325cbef7d4f53d50b37fd4f7421da0ceb89f37bd

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2856/

kbr- commented on test/pylib/scylla_cluster.py:
```diff
@@ -560,17 +563,19 @@ def server_log_filename(self) -> Optional[pathlib.Path]:
 
-    def running_servers(self) -> List[str]:
-        return list(set(self.running.keys()) - self.removed)
+    def running_servers(self) -> List[Tuple[ServerNum, IPAddress]]:
```
Using `ServerInfo` here doesn't work?

alecco replied:
It's only used to return a JSON via API. The type doesn't help.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -295,19 +296,16 @@ async def cql_is_up(self) -> bool:
-        except aiohttp.ClientConnectionError:
+            self.host_id = await api.get_host_id(self.hostname)
+            return True
+        except (aiohttp.ClientConnectorError, RuntimeError):
```
I can use a custom exception. Is this OK?

kbr- replied:
I don't know. I need to understand how Scylla and this code will behave when there's a problem and `/storage_service/hostid/local` throws an exception. We cannot come up with a correct solution before we understand the code.

Therefore I'd like this questions answered:
What will happen if Scylla has a problem throws an exception from the /storage_service/hostid/local endpoint? How will the code behave right now?

kbr- replied:
We discussed this on Slack.

So it turns out that while Scylla is initializing, trying to fetch `/storage_service/hostid/local` might return `404` - the endpoint wasn't created yet.

We want to:
- retry if the error code is `404`
- fail if the error code is `500`

since `RuntimeError` loses this information, we decided that `RESTClient` should throw a better error type from `_fetch` - which doesn't lose this information.

alecco replied:
Done in https://github.com/scylladb/scylladb/commit/8c39b3b7dc7a3d956837eb340c75e00a5b4589a9

kbr- commented on test/pylib/rest_client.py:
```diff
@@ -5,132 +5,115 @@
+        return None  # Unreached
+
+    async def get(self, resource_uri: str, host: Optional[str] = None, port: Optional[int] = None,
+                  params: Optional[Mapping[str, str]] = None) -> None:
```
Return type should be `Any`

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/8c39b3b7dc7a3d956837eb340c75e00a5b4589a9.

Also added missing `return`.

alecco commented on test/pylib/rest_client.py:
```diff
@@ -5,132 +5,112 @@
+            if response_type is not None:
+                # Return response.text() or response.json()
+                return await getattr(resp, response_type)()
+        return None  # Unreached
```
Fixed in https://github.com/scylladb/scylladb/commit/8c39b3b7dc7a3d956837eb340c75e00a5b4589a9

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2898/

kbr- commented:
Maybe we should use our own custom error type after all:
```
>               raise HTTPError(uri, resp.status, f"{text}, params {params}, json {json}")
E               test.pylib.rest_client.HTTPError: <unprintable HTTPError object>
```

kbr- commented:
```
21:57:57.055 ERROR> Failed to start Scylla server at host 127.90.25.6 in scylla-6: [Errno 104] Connection reset by peer
21:57:57.055 ERROR> Error handling request
Traceback (most recent call last):
  File "/usr/lib64/python3.9/site-packages/aiohttp/web_protocol.py", line 422, in _handle_request
    resp = await self._request_handler(request)
  File "/usr/lib64/python3.9/site-packages/aiohttp/web_app.py", line 499, in _handle
    resp = await handler(request)
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_cluster.py", line 870, in _cluster_server_add
    s_info = await self.cluster.add_server()
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_cluster.py", line 538, in add_server
    await server.install_and_start(self.api)
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_cluster.py", line 164, in install_and_start
    await self.start(api)
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_cluster.py", line 345, in start
    if hasattr(self, "host_id") or await self.get_host_id(api):
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_cluster.py", line 299, in get_host_id
    self.host_id = await api.get_host_id(self.ip_addr)
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/rest_client.py", line 137, in get_host_id
    host_uuid = await self.client.get_text("/storage_service/hostid/local", host=server_ip)
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/rest_client.py", line 78, in get_text
    ret = await self._fetch("GET", resource_uri, response_type = "text", host = host,
  File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/rest_client.py", line 60, in _fetch
    async with request(method, uri,
  File "/usr/lib64/python3.9/site-packages/aiohttp/client.py", line 1161, in __aenter__
    self._resp = await self._coro
  File "/usr/lib64/python3.9/site-packages/aiohttp/client.py", line 544, in _request
    await resp.start(conn)
  File "/usr/lib64/python3.9/site-packages/aiohttp/client_reqrep.py", line 890, in start
    message, payload = await self._protocol.read()  # type: ignore
  File "/usr/lib64/python3.9/site-packages/aiohttp/streams.py", line 604, in read
    await self._waiter
aiohttp.client_exceptions.ClientOSError: [Errno 104] Connection reset by peer
```

kbr- commented:
> Maybe we should use our own custom error type after all:

Ah, now I looked at the code, you did introduce a custom type.
(I recall you proposed to import it from the standard library)

But why is it unprintable? You do define `__str__`:
```python
class HTTPError(Exception):
    def __init__(self, uri, code, message):            
        super().__init__(message)
        self.uri = uri
        self.code = code
        self.message = message

    def __str__(self):
        return f"HTTP error {self.status}: {self.message}"
```

kbr- commented:
@alecco let's catch `ClientConnectionError` like before - it's a base for both `ClientConnectorError` and `ClientOSError` (and others).

(plus HTTPError 404 of course)

alecco commented:
> @alecco let's catch `ClientConnectionError` like before - it's a base for both `ClientConnectorError` and `ClientOSError` (and others).
> 
> (plus HTTPError 404 of course)

you mean `ClientError`, right?

kbr- commented:
> you mean ClientError, right?

let's go with `ClientConnectionError`, it's battle-tested and will make the diff smaller (this exception was used before in this code). Also, looking at the docs, `ClientError` has many subexceptions which we would not want to catch (e.g. `ClientPayloadError` - if we get this, we probably want to fail the test).

We know that `ClientConnectionError` works well enough, why change it now.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/3fb77dce37a65733c75ac4b92a0898e8585796db:

- Catch more general HTTP client error `ClientError` (including `ClientOSError`)
- Fix `HTTPError.__str__()`

kbr- commented:
So you decided on `ClientError` after all. It seems like a change for change's sake, there was nothing wrong with `ClientConnectionError`, it didn't give us any problems before.

But fine.

kbr- commented on test/pylib/scylla_cluster.py:
```diff
@@ -295,19 +296,18 @@ async def cql_is_up(self) -> bool:
-        except aiohttp.ClientConnectionError:
+            self.host_id = await api.get_host_id(self.hostname)
+            return True
+        except (aiohttp.ClientError, HTTPError, RuntimeError) as exc:
```
I thought we agreed to **replace** `RuntimeError` with `HTTPError`, not catch them both. `RuntimeError`s should pass through.

I hope it should no longer matter though, because we no longer throw `RuntimeError` from `get_host_id` (IIUC).

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/88d45d24e55c1ac231e596be0fa3284984b9a858

kbr- commented on test/pylib/manager_client.py:
```diff
@@ -190,3 +199,15 @@ async def host_is_down():
+
+    async def get_host_id(self, server_id: ServerNum) -> HostID:
+        """Get host id of a server"""
+        return HostID(await self.client.get_text(f"/cluster/host-id/{server_id}"))
```
The previous version of this PR had `try`+`except`+reraise with additional message.
This was consistent with the function right above it, which still does that.

The code should be written in a consistent manner. The two functions have nearly identical logic, so they should be structured the same way.

I don't get why you decided to restructure it in this version. You did not give any motivation for it in your comments, so I'm guessing the change is unnecessary. It only prolongs the review process. I think we both want this review to end and the PR to get merged, so why?

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/88d45d24e55c1ac231e596be0fa3284984b9a858

kbr- reviewed (approved):
I don't see any major problems, so LGTM.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/88d45d24e55c1ac231e596be0fa3284984b9a858:

- `ScyllaServer.get_host_id()` catch `aiohttp.ClientConnectionError` (to cover `ClientOSError`) and `HTTPError`, removed `RuntimeError`
- `ManagerClient.get_host_id()` reintroduce catch exception and raise with more info

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2931/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/700054abee7ef0dc5dd977b5c4f9643277324af7:

- rebased on latest `master`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2943/

kbr- commented:
Failure due to https://github.com/scylladb/scylladb/issues/8079#issuecomment-1300377785.
Restarting CI

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2950/

nyh commented:
I was about to merge this series when Kamil beat me by two minutes :-)

I just want to note that while I think this series is fine and good and everything, I have somewhat of a problem with the title "test.py" prefix given to all these patches -
I think we completely lost track of what is "part of test.py" - the tool that runs all the tests of many different types - and what is just convenience functions used by only a minority of the tests.
It is arguable whether all this code of tracking Scylla clusters belongs in test.py or it really should be a separate library with a different name (and it really is - it's in test/pylib and not in test.py...) - but I find it hard to justify why stuff like test/pylib/random_tables.py there, doing some highly specific CQL-only stuff. Do we also consider that code "part of test.py"? Why?

alecco commented:
> I just want to note that while I think this series is fine and good and everything, I have somewhat of a problem with the title "test.py" prefix given to all these patches - I think we completely lost track of what is "part of test.py" - the tool that runs all the tests of many different types 

I was just using `test.py` as a top level module (sort of). What could be an appropriate name? `pytest`?
